### PR TITLE
ARROW-3381: [C++] Add bz2 codec

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
AppVeyor build here: https://ci.appveyor.com/project/pitrou/arrow/builds/19685443

**Comment by pitrou** [comment]
The R build failure is unrelated (see ARROW-3593).

File: cpp/src/arrow/util/compression_bz2.cc [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+                    uint8_t* output, int64_t* bytes_read, int64_t* bytes_written,
+                    bool* need_more_output) override {
+    stream_.next_in = const_cast<char*>(reinterpret_cast<const char*>(input));
+    stream_.avail_in = static_cast<unsigned int>(std::min(input_len, kSizeLimit));
+    stream_.next_out = reinterpret_cast<char*>(output);
+    stream_.avail_out = static_cast<unsigned int>(std::min(output_len, kSizeLimit));
+    int ret;
+
+    ret = BZ2_bzDecompress(&stream_);
+    if (ret == BZ_OK || BZ_STREAM_END) {
```
Comment (kou) [inline cpp/src/arrow/util/compression_bz2.cc]: Is `ret ==` missing for `BZ_STREAM_END`?

**Review by kou — CHANGES_REQUESTED** [review]
+1
I have one comment. Can you check it?

File: cpp/src/arrow/util/compression_bz2.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/compression_bz2.cc]: Good catch indeed!

**Comment by pitrou** [comment]
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/19748138

**Comment by pitrou** [comment]
R failure is unrelated. Will merge.

**Pull request closed without merging**